### DISCUSSION TITLE
How to speed XLA code...a challenge

### DISCUSSION THREAD

**Discussion by jecampagne** [jax-ml/jax#11324] [Q&A]
Hi, here a snippet

```python
import jax
from jax import jit
import jax.numpy as jnp

@functools.partial(jit, static_argnums=0)
def complicated_a_func(a_obj, x):
  return a_obj.val * x

class A():
    def __init__(self,val):
        self.val = val
    
    func = complicated_a_func

class A_factory():
    done = False
    _ws = {}
    @classmethod
    def make(cls, n_data=1_000):
        if not A_factory.done:
           A_factory._ws = {'data': [A(val) for val in np.random.uniform(0., 10, size=n_data)]}
            A_factory.done = True
        return A_factory._ws

list_of_A =A_factory.make()['data']        # Build the list

@jit
def f(p):
    list_of_A = A_factory.make()['data']
    data = jnp.array(jax.tree_map(lambda a: a.func(p), list_of_A[:100] ))
    return data
```

Notice that 
1) the A_factory is here quite simple, but in my real use-case it deals with dataset reading.
2) the complicated_a_func is here very simple but can in real use case perform linear algebra manipulation.

The problem is the JITization of the f-function that depends linearly of the size of the fraction of "list_of_A" used in the "jax.tree_map". 

Does one see another more effective way to proceed?
Thanks

**Comment by YouJiacheng** [comment]
1. Don't JIT `f`, only JIT `complicated_a_func`
2. Don't mark `a_obj` as static, but make `A` a custom Pytree to avoid recompiling.

**Reply by jecampagne** [reply]
Thanks @YouJiacheng for your interest, as I am not a JAX expert, can you elaborate a the PyTree trick you suggest?

**Reply by YouJiacheng** [reply]
You can check this first: https://jax.readthedocs.io/en/latest/pytrees.html#extending-pytrees
Note: `aux_data` in pytree is static and will cause recompiling. I think the most of attributes of `a_obj` are arrays, and could be children.

**Reply by jecampagne** [reply]
Hello @YouJiacheng 

I come with this (possible) solution but not yet complete

```python 
import jax
from jax import jit
import jax.numpy as jnp
import numpy as np
from jax.tree_util import register_pytree_node_class

@jit
def complicated_a_func(a_obj, x):
  return a_obj.val * x

@register_pytree_node_class
class A():
    def __init__(self,val):
        if not (type(val) is object or val is None or isinstance(val, A)):
            val = jnp.asarray(val)
        self.val = val
    
    def tree_flatten(self):
      children = (self.val,)
      aux_data = None
      return (children, aux_data)

    @classmethod
    def tree_unflatten(cls, aux_data, children):
      return cls(*children)
      
    func = complicated_a_func

#####
class A_factory():
    done = False
    _ws = {}
    @classmethod
    def make(cls, n_data=1_000):
        if not A_factory.done:
            A_factory._ws = {'data': [A(val) for val in np.random.uniform(0., 10, size=n_data)]}
            A_factory.done = True
        return A_factory._ws

list_of_A =A_factory.make(50_000)['data']
#######
```
Now if I use this function 
```python
def f1(p):
    list_of_A = A_factory.make()['data']
    data = jnp.array(jax.tree_map(lambda a: a.func(p), list_of_A[:100] ))
    return data
```
There is an error
```
      1 def f1(p):
      2     list_of_A = A_factory.make()['data']
----> 3     data = jnp.array(jax.tree_map(lambda a: a.func(p), list_of_A[:100] ))
      4     return data

AttributeError: 'DeviceArray' object has no attribute 'func'
```

Then if I use `jax.lax.map` instead of the `jax.tree_map`, I also get an error
```
-----> jax.lax.map(lambda a: a.func(10.), list_of_A[:100] )
IndexError: tuple index out of range
```

If now I simply use
```python
def f(n, p):
  res = jnp.stack([a.func(10.) for a in list_of_A[:n]])
  return res
```
Then at least i works but the timing seems very long
```python
%time a=f(50_000, 10.)
```
gives even after done twice (ie. XLA compilation should have been done ?)
```
CPU times: user 6.88 s, sys: 109 ms, total: 6.99 s
Wall time: 6.92 s
```

**Reply by YouJiacheng** [reply]
You should pass a proper `is_leaf` predicate to `tree_map`, such that your `A` won't be flattened.

**Reply by YouJiacheng** [reply]
You can print inside `complicated_a_func` to check whether it is compiled more than once.

**Reply by jecampagne** [reply]
Oups! 
1) what would be a "proper" `is_leaf`???
2) I have
```python
@jit
def complicated_a_func(a_obj, x):
  print("compiled...")
  return a_obj.val * x  
.....

def f(n, p):
  res = jnp.stack([a.func(10.) for a in list_of_A[:n]])
  return res
```
and the msg "compiled..." appears only once.
```
%time a=f(1_000, 10.)
compiled...
CPU times: user 157 ms, sys: 7 ms, total: 164 ms
Wall time: 158 ms

%time a=f(1_000, 10.)
CPU times: user 135 ms, sys: 616 µs, total: 136 ms
Wall time: 139 ms

%time a=f(5_000, 10.)
CPU times: user 945 ms, sys: 46.7 ms, total: 992 ms
Wall time: 975 ms

**Reply by YouJiacheng** [reply]
1. e.g. `lambda x: isinstance(x, A)` But just use list comprehension in your case is okay.
2. So compilation time is not the problem. The slowness might be caused by `stack`.

**Reply by jecampagne** [reply]
Great! 
Now do you see a possible more effective way in place of the jnp.stack?

**Reply by YouJiacheng** [reply]
That will depend on your usecase. Do you really need to stack the whole dataset instead of stacking inside batches?

**Reply by jecampagne** [reply]
In fact I only need a 1D vector conatining the results of the  `[a.func(x) for a in list_of_A]`

**Reply by jecampagne** [reply]
Ha an other thing, if in the A object I have a variable which is a number (ie. self.a = 2) then do this variable should go to `aux_data`?

**Reply by YouJiacheng** [reply]
If this variable `children` doesn't break your code, then you don't need to place it at `aux_data`.
i.e. you don't use if for python control flow, array shape, etc.

**Reply by jecampagne** [reply]
Ok, I remark that I need
```python
@jit
def complicated_a_func(a_obj, x):
  return a.obj.K(x)
```
with 
```python
class A():
    def __init__(self,K):
        self.K = K
....
```
that is to say one of the object A arguments is a auxilary function (eg. a kernel). How I show proceed in this case?

And to be more exhaustive, the A object is created with 
a K function,  several ndarrays, two  booleans, 1 flaot, 1int and 1  dictionary containing a flaot and a ndarray.

**Reply by YouJiacheng** [reply]
If all functions are parameterized, you can wrap it with `jax.tree_util.Partial` which is a pytree and make it go to `children`. 
i.e. `k = Partial(k_need_param, param=p)`.
If all functions are non-parameterized, let it go to `aux_data` is okay. But that will trigger re-compilation for each different K.
In mixed case, just like all parameterized case, you can wrap non-parameterized kernel with no-op `Partial(k_non_param)`.

**Reply by jecampagne** [reply]
Well , here the archetype of use-case
```python
class A():
def __init__(self, 
kernel, ### a funcction  
    idx,  ####  int
    flag, ###  bool
       var, ###  float
    paramK, ### ndarray
    params, ### ndarray
      )
....
func = complicated_a_func
```
and `kernel` is a function that depend on `paramK` while
```
def complicated_a_func(a_obj, x):
```
depends on `kernel` (so also on `paramK`) , as well all the others `idx, `flag`, `var`, `params`.

**Reply by YouJiacheng** [reply]
Well, it is okay to let `kernel` go to `aux_data` since it doesn't contain parameters(but accept them as input), and let `paramK` go to `children`.
But I think it is more elegant to wrap `kernel` and `paramK` together using `jax.tree_util.Partial` and let it go to `children`.
If `idx` and `flag` are used in python control flow, they should go to `aux_data`.
`var` and `params` can safely go to `children`.

**Comment by jecampagne** [comment]
@YouJiacheng  
I have a new version based on your advises and I get improvements ! Many thanks. 

I would like to get the last point. 
Here is a short version of my "A" class 
```python
@register_pytree_node_class
class A:
    def __init__(self,
                 kernel, #: Callable[[jnp.ndarray, jnp.ndarray, Dict[str, jnp.ndarray], float, float],jnp.ndarray]
                 kernel_hat, # dict
                 ):
        self.kernel = kernel
       self.kernel_hat = kernel_hat

    def tree_flatten(self):
        children = (self.kernel_hat,    # dict
                    )
        aux_data = {'kernel':self.kernel}
        return (children, aux_data)

    @classmethod
    def tree_unflatten(cls, aux_data, children):
        kernel = aux_data['kernel']
        return cls(kernel, *children)
```
I imagine that you invite me to manage
```
class A:
    def __init__(self,k, <other args>):
         self.k = k
         .... code to deal with other args

    def tree_flatten(self):
        children = (self.k, self.<other args>)
        aux_data = None
        return (children, aux_data)

    @classmethod
    def tree_unflatten(cls, aux_data, children):
        return cls(*children)
```
but then in my A_factory is it correct then to do
```python
a = A(Partial(kernel, params=kernal_hat), <other args>)
```
at least it seems to work...

**Comment by jecampagne** [comment]
Many thanks to @YouJiacheng, even if I do not master (yet) the PyTree (ie. still complicated to see what should go to children and to aux_data) is doing very well. We can close this thread.

**Comment by jecampagne** [comment]
Hi @YouJiacheng 
With your kind help I come with the following schema (here simplifyied on purpose)

```python
import jax
from jax import jit
import jax.numpy as jnp
import numpy as np

@jit
def K(x):
  return x**2

@jit
def complicated_a_func(a_obj, x):
  print("compiled...")
  return a_obj.val * a_obj.f(x)

@register_pytree_node_class
class A():
    def __init__(self,val, f):
        if not (type(val) is object or val is None or isinstance(val, A)):
            val = jnp.asarray(val)
        self.val = val

        self.f = f

    def __repr__(self):
      return "A(val={})".format(self.val)

    
    def tree_flatten(self):
      children = (self.val,self.f)
      aux_data = None
      return (children, aux_data)

    @classmethod
    def tree_unflatten(cls, aux_data, children):
      return cls(*children)

    func = complicated_a_func


class A_factory():
    done = False
    _ws = {}
    @classmethod
    def make(cls, n_data=5_000):
        if not A_factory.done:
            A_factory._ws = {'data': [A(val, Partial(K)) for val in np.random.uniform(0., 10, size=n_data)]}
            A_factory.done = True
        return A_factory._ws
        
list_of_A =A_factory.make()['data']


def f50(p):
    list_of_A = A_factory.make()['data']
    data = jnp.array([a.func(p) for a in list_of_A[:5_000]])
    return data

def f50b(p):
    list_of_A = A_factory.make()['data']
    data = jnp.array(jax.tree_map(lambda a: a.func(p), list_of_A[:5_000], is_leaf=lambda x: isinstance(x, A) ))
    return data
```

The point is that `f50` or `f50b` takes about 2sec on a K80 Google Colab GPU
```
%time tmp=f50b(10.)
%timeit tmp=f50b(10.)
compiled...
CPU times: user 2.17 s, sys: 613 ms, total: 2.78 s
Wall time: 2.02 s
1 loop, best of 5: 1.91 s per loop
```
but here it is just 5000 eveluations of "a * x**2" with "a" an random number [0;10] and x a real (=10.)
  with even with a python for-loop would takes 30 micro-secondes or so.

My question is: what is the mechanism that delay so much the computation ?

**Reply by YouJiacheng** [reply]
1. When there is `DeviceArray` in list, convert a large list to a `jnp.ndarray` is very slow, much slower than `numpy`.
2. dispatch overhead of JAX is relatively high.

**Reply by YouJiacheng** [reply]
Also, you transfer data from CPU to GPU in a one number per transfer manner, which is inefficient.

**Reply by YouJiacheng** [reply]
I suggest you can group data by their kernel(un-parameterized version), convert them to array by `numpy`, then use JAX (with `vmap`).

**Reply by jecampagne** [reply]
Well, 

concerning point 1.
```python
a = [x for x in np.random.uniform(0., 10, size=5000)]
```
and 
```python
%time
x=10.
res = jnp.array([v*x**2 for v in a])
```
CPU times: user 4 µs, sys: 12 µs, total: 16 µs
Wall time: 32.9 µs

Now concerning  "CPU->GPU", I test on CPU Google Colab
```
%time tmp=f50b(10.)
%timeit tmp=f50b(10.)
```
leads to
```
compiled...
CPU times: user 1.17 s, sys: 58.3 ms, total: 1.22 s
Wall time: 1.84 s
1 loop, best of 5: 832 ms per loop
```

It is surprising because at the reason I open  this Q&A thread in fact is that the XLA compilation was very long in my use case (400 sec or so) but then the execution was few mili-sec). 
Now the compilation last 8sec which is great, but the execution is 1sec or so. 

I am a bit puzzled

**Reply by YouJiacheng** [reply]
```python
from contextlib import contextmanager
@contextmanager
def timer():
    from time import perf_counter
    t = perf_counter()
    yield
    print(perf_counter() - t)

import jax.numpy as jnp
import jax
import numpy as np

a = jax.block_until_ready([jnp.array(x) for x in np.random.uniform(0., 10, size=5000)])
with timer():
    jnp.array(a).block_until_ready() # 1.6991751063615084s

with timer():
    jax.jit(jnp.array)(a).block_until_ready() # 11.288538832217455s

with timer():
    jax.jit(jnp.array)(a).block_until_ready() # 0.05233551189303398s

b = list(np.random.uniform(0., 10, size=5000))
with timer():
    jnp.array(b).block_until_ready() # 0.016881123185157776s

# NOTE: following 2 testcases should run separately, since JAX will cache CPU array for DeviceArray
with timer():
    jnp.array(np.array(a)).block_until_ready() # 0.2592076510190964s

with timer():
    jnp.array([np.array(x) for x in a]).block_until_ready() # 0.27834889478981495s
```

**Reply by jecampagne** [reply]
> I suggest you can group data by their kernel(un-parameterized version), convert them to array by `numpy`, then use JAX (with `vmap`).

Well I would like of course test your suggestion but I wander if I understand correctly what does it means in context. 
A the end what is matter is for the user is 
```python
def f(p):
   list_of_A = A_factory.make()['data']
   res = jnp.array([a.func(p) for a in list_of_A])
  return res
```

**Reply by YouJiacheng** [reply]
You can use a workaround: `jnp.array([np.array(a.func(p)) for a in list_of_A])`, if you must combine the list(I don't know why you need to stack all data into a single array).
If there are many flags can affect control flow, my suggestion is invalid, if not:
```python
vals = np.random.uniform(0., 10, size=5000) # group of data with same control flow(including kernel)
a = A(vals, Partial(K, paramK_stacked))
data['group_tag'] = a
res = jnp.concatenate([jax.vmap(A.func, in_axes=(0, None))(a, p) for a in data.values()])
```

**Reply by jecampagne** [reply]
In fact the res is used like that
```python
res1 = jnp.array([x.func(p) for x in list_of_A])   # shape N1
res2 = jnp.array([b.func(p) for x in list_of_B]) # shape (N1*N2)
res2 = res2.reshape(N1,N2)
res = res1[:,jnp.newaxis] * res2   # shape N1,N2
```

**Reply by YouJiacheng** [reply]
I think you can simply use `jnp.array([np.array(a.func(p)) for a in list_of_A])` if `N1 * N2` less than 1 billion.

**Reply by jecampagne** [reply]
This looks strange to use `np.array(a.func(p))` inside the list completion.

**Reply by YouJiacheng** [reply]
`jnp.array([np.array(a.func(p)) for a in list_of_A])` or `jnp.array(np.array([a.func(p) for a in list_of_A]))` doesn't make much difference. The idea is to stack arrays using `numpy`(much faster than `lax` or `jnp` without JIT which also need compilation), at the cost of data transfer.

**Reply by YouJiacheng** [reply]
```python
from contextlib import contextmanager
@contextmanager
def timer():
    from time import perf_counter
    t = perf_counter()
    yield
    print(perf_counter() - t)

import jax.numpy as jnp
import jax
import numpy as np

a = jax.block_until_ready([jnp.array(x)[None] for x in np.random.uniform(0., 10, size=5000)])
with timer():
    jnp.array(a).block_until_ready() # 1.8s, tree concat compile faster

with timer():
    jnp.array(a).block_until_ready() # 1.3s, tree concat execute slower because of more operations and intermediate

with timer():
    jax.lax.concatenate(a, 0).block_until_ready() # 8.3s, direct concat compile slower

with timer():
    jax.lax.concatenate(a, 0).block_until_ready() # 0.050s, direct concat execute faster
```

**Reply by jecampagne** [reply]
That"s amazing ! Thanks

**Reply by jecampagne** [reply]
Hi @YouJiacheng 

1) The use of `np.array`seems to break the autodiff...

```python
a = [x for x in np.random.uniform(0., 10, size=5000)]
def f(x):
    return jnp.array(np.array([v*x**2 for v in a]))

jax.grad(f)(10.)
```
leads to
```
TracerArrayConversionError: The numpy.ndarray conversion method __array__() was called on the JAX Tracer object Traced<ConcreteArray(170.93002319335938, dtype=float32)>with<JVPTrace(level=2/0)> with
  primal = DeviceArray(170.93002, dtype=float32)
  tangent = Traced<ShapedArray(float32[])>with<JaxprTrace(level=1/0)> with
    pval = (ShapedArray(float32[]), None)
    recipe = JaxprEqnRecipe(eqn_id=<object object at 0x7f0afb9f7e50>, in_tracers=(Traced<ConcreteArray(1.70930016040802, dtype=float32):JaxprTrace(level=1/0)>, Traced<ShapedArray(float32[], weak_type=True):JaxprTrace(level=1/0)>), out_tracer_refs=[<weakref at 0x7f0afb0c84f0; to 'JaxprTracer' at 0x7f0afb0c8450>], out_avals=[ShapedArray(float32[])], primitive=xla_call, params={'device': None, 'backend': None, 'name': 'jvp(fn)', 'donated_invars': (False, False), 'inline': True, 'keep_unused': False, 'call_jaxpr': { lambda ; a:f32[] b:f32[]. let
    c:f32[] = convert_element_type[new_dtype=float32 weak_type=False] b
    d:f32[] = mul c a
  in (d,) }}, effects=set(), source_info=SourceInfo(traceback=<jaxlib.xla_extension.Traceback object at 0x7f0afb08bc30>, name_stack=NameStack(stack=(Transform(name='jvp'),))))
```

2) A point related to your first answer, it  was
```python
@jit
def f(p):
    list_of_A = A_factory.make()['data']
    data = jnp.array(jax.tree_map(lambda a: a.func(p), list_of_A[:100] ))
    return data
```
> Don't JIT f, only JIT complicated_a_func

can you elaborate a little bit why the `@jit`of `f` leads to a very XLA compilation time? Does it trigger the compillation of "func" for every element of   `list_of_A[:100]` ?

**Reply by YouJiacheng** [reply]
point 1: yes if you use `numpy` function(except constant, or non-array op e.g. `finfo`), then you can't do any transform.
point 2: actually worse, JAX will inline all call of `func`, produce extremely long jaxpr.

**Comment by YouJiacheng** [comment]
Your test for point 1 is specially optimized by JAX: If all items are not `DeviceArray`, `jnp.array` will use `np.array` first.
https://github.com/google/jax/blob/118db407f24f20a77ef491d504290f8c29d57d05/jax/_src/numpy/lax_numpy.py#L1868-L1890

If there is any `DeviceArray`, `jnp.array` will use `jnp.stack`, which use `jnp.concatenate`, which is very slow in op-by-op code (i.e. non-JIT) for large list even after workaround. (8s -> 1.7s)
https://github.com/google/jax/blob/118db407f24f20a77ef491d504290f8c29d57d05/jax/_src/numpy/lax_numpy.py#L1661-L1685

**Comment by jecampagne** [comment]
Hello @YouJiacheng 

I found interesting the discussion #5322 and got an idea
```python
def pytrees_stack(pytrees, axis=0):
    results = jax.tree_util.tree_map(
        lambda *values: jnp.stack(values, axis=axis), *pytrees)
    return results

@jit
def test(p):
  data = jax.vmap(lambda x: complicated_a_func(x,p))(pytrees_stack(list_of_A))
  return data
```
Notice that 1) I do not use the "func" function of the `A` object but directly the external `complicated_a_func`; 2) I use the jit decorator.

Then
```python
with timer():
  test(10.)
with timer():
  test(10.)
```
gives
```
compiled...
2.3561921679997795
0.00023758199995427276
```
and I can use it in the following expression
```python
jax.grad(lambda p: jnp.sum(test(p)))(10.)
jax.vmap(lambda p: jnp.sum(test(p)))(jnp.array([10., 20.]))
```
what do you think?

**Reply by YouJiacheng** [reply]
That is "I suggest you can group data by their kernel(un-parameterized version), convert them to array by numpy, then use JAX (with vmap)."
If A has different `aux_data`(i.e. kernel, flag, etc,), `tree_map(f, *list_of_A)` will produce an error. And logically you can't vmap on data points with different python-control flow.
Alternately, you can use `lax.switch`(be careful, it will be converted to `lax.select` under `vmap`) and use index to choose kernel.

**Reply by jecampagne** [reply]
Ok  @YouJiacheng, do not worry,  your suggestion comes too early for my understanding. Sorry. 
Now if everything is in the "children" tuple of the PyTrees (A object) then it should be ok?
